Extractor/TypeCollectors/TypeManager.cs

```diff
@@ -150,16 +156,61 @@ private async Task ReadTypeHiearchies(ITypeAndNodeSource source, CancellationTok
             if (config.Events.Enabled && !eventTypesRead)
             {
                 mask |= (uint)NodeClass.ObjectType;
-                rootNodes.Add(ObjectTypeIds.BaseEventType);
+                // Avoid adding the BaseEventType node if we are also adding the BaseObjectType node, since
+                // it's a parent. The browser doesn't really like if you have loops, though it will work.
+                if (!fdmEnabled) rootNodes.Add(ObjectTypeIds.BaseEventType);
                 eventTypesRead = true;
                 log.LogInformation("Loading event type hierarchy");
             }
+            if ((fdmEnabled || config.Extraction.Relationships.Enabled) && !referenceTypesRead)
+            {
+                mask |= (uint)NodeClass.ReferenceType;
+                rootNodes.Add(ReferenceTypeIds.References);
+                referenceTypesRead = true;
+                log.LogInformation("Loading reference type hierarchy");
+            }
+            if (fdmEnabled && !typeDefsRead)
+            {
+                mask |= (uint)NodeClass.VariableType | (uint)NodeClass.ObjectType;
+                rootNodes.Add(ObjectTypeIds.BaseObjectType);
+                rootNodes.Add(VariableTypeIds.BaseVariableType);
+                typeDefsRead = true;
+                log.LogInformation("Loading object type and variable type hierarchies");
+            }
+            if (fdmEnabled)
+            {
+                referenceMode = HierarchicalReferenceMode.Forward;
+                loadReferences = true;
+            }
+
             if (!rootNodes.Any()) return;
 
-            var result = await source.LoadNodes(rootNodes, mask, HierarchicalReferenceMode.Disabled, "the type hierarchy", token);
+
+            var result = await source.LoadNodes(rootNodes, mask, referenceMode, "the type hierarchy", token);
+
             foreach (var node in result.Nodes)
             {
-                NodeMap[node.Id] = node;
+                NodeMap.TryAdd(node.Id, node);
+                // Any nodes discovered here will be in the type hierarchy.
+                node.IsChildOfType = true;
+            }
+
+            if (loadReferences)
```

Comment:
I'd like to keep it separate, just to be explicit about what's going on. If this was just `if (fdmEnabled)` it would be less clear what it was actually doing.

Reply:
Maybe a comment could work as well, but this is fine